### PR TITLE
Add support for custom types in arrays and tuples

### PR DESCRIPTION
**Pull request by digorithm**
Closes https://github.com/FuelLabs/fuels-rs/issues/306. 

You can find more context in https://github.com/FuelLabs/fuels-rs/issues/306. The gist is that, recently, there were many critical changes to the JSON ABI outputted by the compiler w.r.t arrays and tuples with custom types in them. 

This PR adds support for these changes; Making it possible to use arrays with custom types, e.g., `[Person; 2]` and `[State; 4]`, and tuples with custom types, e.g., `(u64, Person)` and `(u64, State)`.

This touches virtually everything related to the custom types generation in the abigen and the function selector builder. 

Here are some examples of what's possible now:

```Rust
// Tuple with struct.
let my_struct_tuple = (
    42,
    Person {
        name: "Jane".to_string(),
    },
);
let response = instance
    .returns_struct_in_tuple(my_struct_tuple.clone())
    .call()
    .await
    .unwrap();

assert_eq!(response.value, my_struct_tuple);

// Tuple with enum.
let my_enum_tuple: (u64, State) = (42, State::A());

let response = instance
    .returns_enum_in_tuple(my_enum_tuple.clone())
    .call()
    .await
    .unwrap();

assert_eq!(response.value, my_enum_tuple);
```

And with arrays:

```Rust
let persons = vec![
    Person {
        name: "John".to_string(),
    },
    Person {
        name: "Jane".to_string(),
    },
];

let result = contract_instance
    .array_of_structs(persons)
    .call()
    .await
    .unwrap();

assert_eq!("John", result.value[0].name);
assert_eq!("Jane", result.value[1].name);

let states = vec![State::A(), State::B()];

let result = contract_instance
    .array_of_enums(states.clone())
    .call()
    .await
    .unwrap();

assert_eq!(states[0], result.value[0]);
assert_eq!(states[1], result.value[1]);
```

### CODE REVIEW

**Review by MujkicA — CHANGES_REQUESTED** [review]

File: packages/fuels-core/src/json_abi.rs [hunk 1]
```diff
@@ -491,25 +491,57 @@ impl ABIParser {
...
+                // "[struct Person; 2]" and we want to extract the
+                // length, which in this example is 2.
+                let array_length: String = array_type_field.split(';').collect::<Vec<&str>>()[1]
+                    .trim()
+                    .chars()
+                    .filter(|c| c.is_digit(10))
+                    .collect();
+
+                result.push(';');
+                result.push(array_length.chars().next().unwrap());
```
[the remark this answers is not in the document]
Comment (digorithm) [reply packages/fuels-core/src/json_abi.rs]: Excellent catch, improved the logic here, that `.chars()` was a mistake!

**Comment by digorithm** [comment]
Great review @MujkicA! I bundled all changes related to your review in https://github.com/FuelLabs/fuels-rs/pull/356/commits/b1c399d3123969858e890b4974f58704b1053ea9.

**Review by MujkicA — APPROVED** [review]
LGTM

**Review by hal3e — APPROVED** [review]

**Review by vnepveu — COMMENTED** [review]
Just catching up, great work!

**Pull request merged**